### ISSUE TITLE
Performance Issue: GetSiteSettings

### ISSUE THREAD

**Issue by MatteoPiovanelli** [OrchardCMS/Orchard#8284]
Tagging @sebastienros @HermesSbicego-Laser @BenedekFarkas 

We have been doing some performance work on our Orchard installations lately.
As part of that I've been profiling a copy of one of our tenants that I have on my dev box.
The process I was profiling was the repeated access (a few hundred times) to a cached page (a version from my local tenant of https://www.anfiteatromorenicoivrea.it/erbaluce-di-caluso).

Roughly 48% CPU time is used in Orchard.Core.Settings.Service.SiteService.GetSiteSettings() on this line:

    return _contentManager.Get<ISite>(siteId, VersionOptions.Published);

That is of course a `WorkContext.CurrentSite` being called.

Going deeper, I found that the line where that gets called most times "uses" about 51% CPU time and is the following:

    var sslSettings = _wca.GetContext().CurrentSite.As<SslSettingsPart>();

From Orchard.SecureSocketsLayer.Services.StrictTransportSecurityMiddlewareProvider.GetOwinMiddlewares().

My first instinct here would be to cache that `CurrentSite`, but I would not feel 100% positive on when to evict that cache because of updates to any setting. A simpler fix would be to cache the `SslSettingsPart`: we would easily know to evict that with an handler on that ContentPart.

Would anyone validate that these results are not artifacts of how my box and my tenant are set up?
Do you see any drawback to the approach I propose as a possible solution?

On further testing on pages that are not cached, the "contribuition" of the call to `GetSiteSettings()` to the used CPU seems to drop significantly, to about 1%.

**Comment by MatteoPiovanelli** [comment]
On a similar analysis, we found a similar performance issue with:

    Orchard.Localization.Services.SiteCultureSelector.GetCulture(HttpContextBase context)

It's a different tenant, so a different configuration, and for cached contents that method used roughly 52% CPU time.
(again, based off what VS2017's profiler is showing me)

**Comment by MatteoPiovanelli** [comment]
In non-cached pages we see, for the first tenant, around 26% of CPU cylces are in `Orchard.Core.Navigation.Services.NavigationManager.BuildMenu(IContent menu)`. A good chunk of that is in the Merge and Arrange static methods. I would think that wrapping those in a call to a cache manager would improve stuff. A key could be generated by appending a list of MenuItem ids (sorted?)

**Comment by sebastienros** [comment]
1-

I don't understand how calling `As<SslSettingsPart>` will trigger `Orchard.SecureSocketsLayer.Services.StrictTransportSecurityMiddlewareProvider.GetOwinMiddlewares()`

Can you explain ?

2- I assume this is the same issue as 1- as it calls `_workContextAccessor.GetContext().CurrentSite`

3- In Orchard Core we decided to cache the menu on the front end pages. Even if the page is not cached.

**Comment by MatteoPiovanelli** [comment]
1- I was not clear, sorry. The call to 
`var sslSettings = _wca.GetContext().CurrentSite.As<SslSettingsPart>();` 
is in the `GetOwinMiddlewares()` method. 
I would replace that line with something like (I forget the exact syntax right now)

    var sslSettings = _cacheManager.Get("a proper key", (ctx) => {
        ctx.Monitor(_signals.When("a signal triggered by updates to the SslSettingsPart"));
        return _wca.GetContext().CurrentSite.As<SslSettingsPart>();
    });

Since this would be in the `GetOwinMiddlewares()`, and I don't rightly know what is going on there, I wonder whether there would be some weird repercussions from having stuff being fetched from cache.

2- Pretty similar issue. However, `CurrentCulture` is computed by going through several providers, each giving its "opinion" on what the culture is. In our tests, we found that the expensive one is `SiteCultureSelector`, and it's pretty easy to cache the results of its implementation of `GetCulture(HttpContextBase context)`.

3- The tenant we were running the tests on is relatively menu-heavy. In most cases, I would suspect the optimization/caching of the methods from `NavigationManager` to be less impactful. 

Bonus- For further speedups, what is your opinion on caching stuff in the `XPartDriver.Display` methods? Something along the lines of:
```
protected override DriverResult Display(XPart part, string displayType, dynamic shapeHelper) {
    return ContentShape("Parts_XPart", () => {
        return _cacheManager.Get("this key makes sense", (ctx) => {
            // actual computations for the shape, and perhaps a ctx.Monitor to control eviction
            return shapeHelper.Parts_XPart(parameters);
        });
    });
}
```

**Comment by MatteoPiovanelli** [comment]
I should have a PR with 1, 2 and 3 ready tomorrow

**Comment by MatteoPiovanelli** [comment]
On my tenants/setup
1+2 gave me roughly a 30% reduction of the profiled CPU time for cached pages
3 gave roughly a 40-50% reduction of the profiled CPU time for non-cached pages where it used to rebuild the menu from scratch. 

I am preparing a PR.

**Comment by MatteoPiovanelli** [comment]
PR done. I still would like @sebastienros ' opinion on my bonus question above.

**Comment by sebastienros** [comment]
You can cache the database calls in the shapes, I would still return a new shape as some other handlers might add alternates for instance.

**Comment by MatteoPiovanelli** [comment]
I think we mean the same thing then. I'll start looking for opportunities to do that.

**Comment by strongus** [comment]
I am not sure if this is related but I get very slow admin page loads on our Azure Web App deployment if Orchard.SecureSocketsLayer is enabled and if I disable that module it's like night and day faster. I looked at ApplicationInsights and it the slow processes seem to be coming from this:

StrictTransportSecurityMiddlewareProvider+<<GetOwinMiddlewares>b__6_1>d.MoveNext

It's not so slow on my development machine but on the Azure deployed site it makes it almost unbearably slow to use the Admin side of the site, but it doesn't really seem to be slowing down the front-end anonymous user pages. However, I noticed that it does seem to slow down authenticated pages (that are not in the admin area) too.  

Not sure if this helps but thought I'd chime in since I am having what seems to be a similar problem.
